### PR TITLE
[SMTChecker] Move expression handling to EncodingContext

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #6737

### CODE REVIEW

**Comment by chriseth** [comment]
Pleas rebase.

**Comment by leonardoalt** [comment]
Rebased. This one is a lot shorter

File: libsolidity/formal/EncodingContext.cpp [hunk 1]
```diff
@@ -117,6 +118,43 @@ void EncodingContext::setUnknownValue(SymbolicVariable& _variable)
 	setSymbolicUnknownValue(_variable, m_solver);
 }
 
+/// Expressions
+
+shared_ptr<SymbolicVariable> EncodingContext::expression(solidity::Expression const& _e)
+{
+	if (!knownExpression(_e))
```
Comment (leonardoalt) [inline libsolidity/formal/EncodingContext.cpp:125]: This will eventually be changed to `solAssert` when we have more support.

Comment (leonardoalt) [reply libsolidity/formal/EncodingContext.cpp:125]: Also not changing it now keeps the tests the same for this PR.

File: libsolidity/formal/EncodingContext.cpp [hunk 2]
```diff
@@ -117,6 +118,43 @@ void EncodingContext::setUnknownValue(SymbolicVariable& _variable)
...
+
+bool EncodingContext::createExpression(solidity::Expression const& _e, shared_ptr<SymbolicVariable> _symbVar)
+{
+	solAssert(_e.annotation().type, "");
+	if (knownExpression(_e))
+	{
+		expression(_e)->increaseIndex();
+		return false;
+	}
+	else if (_symbVar)
```
Comment (leonardoalt) [inline libsolidity/formal/EncodingContext.cpp:138]: Needed for two special cases from `SMTChecker`: string literals and global symbols.

File: libsolidity/formal/SMTChecker.cpp [hunk 3]
```diff
@@ -332,19 +331,18 @@ void SMTChecker::endVisit(VariableDeclarationStatement const& _varDecl)
...
 		if (auto init = _varDecl.initialValue())
 		{
-			auto symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_expressions[init]);
-			/// symbTuple == nullptr if it is the return of a non-inlined function call.
-			if (symbTuple)
-			{
-				auto const& components = symbTuple->components();
-				auto const& declarations = _varDecl.declarations();
+			auto symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_context.expression(*init));
+			solAssert(symbTuple, "");
```
Comment (leonardoalt) [inline libsolidity/formal/SMTChecker.cpp:335]: Now the expression of a non-inlined function call is created. If it has more than one value, it is a tuple without components.

**Review by chriseth — APPROVED** [review]

**Pull request merged**